### PR TITLE
feat(*): add prometheus metrics

### PR DESCRIPTION
**Pull request by lurais**
1.add prometheus metrics

**What does this PR do?**
add prometheus metrics
**Why are these changes required?**
to inquire about the system performance

**This PR has been tested by:**
- Unit Tests
- Manual Testing

**Follow up**

**Extra details**

### CODE REVIEW

File: chainbase/src/main/java/org/tron/common/storage/leveldb/LevelDbDataSourceImpl.java [hunk 1]
```diff
@@ -23,22 +23,16 @@
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (halibobo1205) [inline chainbase/src/main/java/org/tron/common/storage/leveldb/LevelDbDataSourceImpl.java:26]: do not merge import *

File: chainbase/src/main/java/org/tron/common/storage/leveldb/LevelDbDataSourceImpl.java [hunk 2]
```diff
@@ -208,31 +206,48 @@ public void setDBName(String name) {
...
     resetDbLock.readLock().lock();
     try {
-      return database.get(key);
+      byte[] result = database.get(key);
+      Metrics.counterInc(MetricKeys.Counter.DB_GET, 1, getEngine(),getDBName(),
+              Objects.isNull(result)?MetricLabels.Counter.DB_GET_MISS:MetricLabels.Counter.DB_GET_SUCCESS);
+      return result;
     } finally {
       resetDbLock.readLock().unlock();
+      Metrics.histogramObserve(requestTimer);
```
Comment (halibobo1205) [inline chainbase/src/main/java/org/tron/common/storage/leveldb/LevelDbDataSourceImpl.java:220]: check style ,spaces are required on both sides of the operator

File: chainbase/src/main/java/org/tron/common/storage/leveldb/LevelDbDataSourceImpl.java [hunk 3]
```diff
@@ -208,31 +206,48 @@ public void setDBName(String name) {
...
+            MetricKeys.Histogram.DB_SERVICE_LATENCY,
+            getEngine(), getDBName(), MetricLabels.Histogram.DB_PUT);
     resetDbLock.readLock().lock();
     try {
       database.put(key, value, writeOptions);
     } finally {
       resetDbLock.readLock().unlock();
+      Metrics.histogramObserve(requestTimer);
+      Metrics.histogramObserve(MetricKeys.Histogram.DB_SERVICE_VALUE_BYTES, ByteUtil.getSize(value),
+              getEngine(),getDBName());
```
Comment (halibobo1205) [inline chainbase/src/main/java/org/tron/common/storage/leveldb/LevelDbDataSourceImpl.java:236]: check style ,spaces are required on both sides of the operator

File: chainbase/src/main/java/org/tron/core/db2/core/SnapshotImpl.java [hunk 4]
```diff
@@ -48,21 +54,36 @@ public byte[] get(byte[] key) {
...
   private byte[] get(Snapshot head, byte[] key) {
+    Histogram.Timer requestTimer = Metrics.histogramStartTimer(
+            MetricKeys.Histogram.SNAPSHOT_SERVICE_LATENCY, root.getDbName(), MetricLabels.Histogram.SNAPSHOT_GET);
     Snapshot snapshot = head;
     Value value;
     if (isOptimized) {
       value = db.get(Key.of(key));
+      if(Objects.nonNull(value)){
+        Metrics.histogramObserve(requestTimer);
+      }
```
Comment (halibobo1205) [inline chainbase/src/main/java/org/tron/core/db2/core/SnapshotImpl.java:65]: if null, not record?

File: chainbase/src/main/java/org/tron/core/db2/core/SnapshotImpl.java [hunk 5]
```diff
@@ -48,21 +54,36 @@ public byte[] get(byte[] key) {
...
       }
 
       snapshot = snapshot.getPrevious();
     }
 
-    return snapshot == null ? null : snapshot.get(key);
+    byte[] result = snapshot == null ? null : snapshot.get(key);
+    Metrics.counterInc(MetricKeys.Counter.SNAPSHOT_GET,1,root.getDbName(),
+            Objects.isNull(result)?MetricLabels.Counter.SNAPSHOT_GET_MISS:MetricLabels.Counter.SNAPSHOT_GET_SUCCESS,
+            MetricLabels.Counter.SNAPSHOT_GET_REACH_ROOT);
```
Comment (halibobo1205) [inline chainbase/src/main/java/org/tron/core/db2/core/SnapshotImpl.java:85]: same as MetricKeys.Counter.DB_GET ?

File: chainbase/src/main/java/org/tron/core/db2/core/SnapshotManager.java [hunk 6]
```diff
@@ -282,6 +290,8 @@ private void refreshOne(Chainbase db) {
     List<Snapshot> snapshots = new ArrayList<>();
 
     SnapshotRoot root = (SnapshotRoot) db.getHead().getRoot();
+    Histogram.Timer flushTimer = Metrics.histogramStartTimer(
+            MetricKeys.Histogram.CHECKPOINT_LATENCY,root.getDbName());
```
Comment (halibobo1205) [inline chainbase/src/main/java/org/tron/core/db2/core/SnapshotManager.java:294]: check style

File: common/src/main/java/org/tron/common/prometheus/MetricsHistogram.java [hunk 7]
```diff
@@ -41,6 +41,14 @@ public class MetricsHistogram {
...
     init(MetricKeys.Histogram.HTTP_BYTES, "http_bytes traffic.",
         "url", "status");
+    init(MetricKeys.Histogram.DB_SERVICE_LATENCY, "db service latency.",
+            "type", "db", "op");
+    init(MetricKeys.Histogram.CHECKPOINT_LATENCY, "checkpoint flush latency.",
+            "type");
+    init(MetricKeys.Histogram.DB_SERVICE_VALUE_BYTES, "db service value bytes.",
+            "type","db");
+    init(MetricKeys.Histogram.SNAPSHOT_SERVICE_VALUE_BYTES, "snapshot value bytes.","db");
+    init(MetricKeys.Histogram.SNAPSHOT_SERVICE_LATENCY,"snapshot service latency","db","op");
```
Comment (halibobo1205) [inline common/src/main/java/org/tron/common/prometheus/MetricsHistogram.java:51]: check style

File: framework/src/main/java/org/tron/program/FullNode.java [hunk 8]
```diff
@@ -141,6 +146,18 @@ public static void main(String[] args) {
     rpcApiService.blockUntilShutdown();
   }
 
+  private static void initPrometheusMetrics(TronApplicationContext context) {
+    try {
```
Comment (halibobo1205) [inline framework/src/main/java/org/tron/program/FullNode.java:150]: move to org.tron.common.prometheus.Metrics?

File: chainbase/src/main/java/org/tron/core/db2/core/SnapshotImpl.java (hunk 4 above)
Comment (lurais) [reply chainbase/src/main/java/org/tron/core/db2/core/SnapshotImpl.java:65]: if null,the result will be recorded when the query from snapshotRoot completed.

Comment (lurais) [reply chainbase/src/main/java/org/tron/core/db2/core/SnapshotImpl.java:65]: > if null, not record?

if null,the result will be recorded when the query from snapshotRoot completed.

File: chainbase/src/main/java/org/tron/core/db2/core/SnapshotImpl.java (hunk 5 above)
Comment (lurais) [reply chainbase/src/main/java/org/tron/core/db2/core/SnapshotImpl.java:85]: > same as MetricKeys.Counter.DB_GET ?

SNAPSHOT_GET is used to measure the snapshotImpl performance while the DB_GET used to measure the snapshotRoot performance.

File: chainbase/src/main/java/org/tron/common/storage/leveldb/LevelDbDataSourceImpl.java (hunk 1 above)
Comment (lurais) [reply chainbase/src/main/java/org/tron/common/storage/leveldb/LevelDbDataSourceImpl.java:26]: > do not merge import *

ok

**Review by halibobo1205 — APPROVED** [review]

**Pull request merged**